### PR TITLE
Support for scoped generators (fixes #14)

### LINKED ISSUE #14: Need to support new npm scoped packages
Initially opened here: https://github.com/yeoman/yo/issues/231  
npm documentation: https://github.com/npm/npm/issues/5239  

This will create module under a different kind of path, e.g.: `/usr/local/lib/node_modules/@foo/generator-bar/`

### PR DESCRIPTION
**Pull request by molant**
This is the easiest way to look for scoped packages. It does not use the scope name when calling yo (so not what @SBoudrias expects in https://github.com/yeoman/yo/issues/231).
I'll see if I can make the other version and then you can decide which one to use.

### CODE REVIEW

**Comment by molant** [comment]
You can invoke now the generator via
```
yo @mycompany/generatorname
```

File: lib/environment.js [hunk 1]
```diff
@@ -150,6 +150,18 @@ Environment.prototype.register = function register(name, namespace) {
...
+  var folders = name.split(path.sep);
+  var master = namespace.split(':')[0];
+  for (var i = 0, l = folders.length; i < l; i++) {
+    if (folders[i].indexOf('generator-' + master) !== -1){
+      break;
+    }
+  }
+
+  if (folders[i - 1].indexOf('@') !== -1) {
+    namespace = folders[i - 1] + '/' + namespace;
```
Comment (SBoudrias) [inline lib/environment.js]: All this logic is not at the right place. It should at least be on top of the error being thrown if no namespace is available

File: lib/environment.js [hunk 2]
```diff
@@ -150,6 +150,18 @@ Environment.prototype.register = function register(name, namespace) {
     return this.error(new Error('Unable to determine namespace.'));
   }
 
+  var folders = name.split(path.sep);
+  var master = namespace.split(':')[0];
+  for (var i = 0, l = folders.length; i < l; i++) {
```
Comment (SBoudrias) [inline lib/environment.js]: So, this for/if/break logic is super hard to follow. Refactor the style to be more functional (check some lodash helpers).

What you want is to find the last occurrence of a folder name starting with a `@` sign. And you want to return this name.

File: lib/resolver.js [hunk 3]
```diff
@@ -63,6 +63,10 @@ resolver.findGeneratorsIn = function (searchPaths) {
 
   searchPaths.forEach(function (root) {
     if (!root) return;
+    modules = glob.sync('@*', {cwd: root, stat: true}).reduce(function(acum, current){
```
Comment (SBoudrias) [inline lib/resolver.js]: This could probably be bundled up into a single glob statement. Also, we don't want this method to be recursive.

This is very performance sensible code as it'll hold any generator (and yo) from running. We want to keep the looping to a minimal.

Maybe something like: (haven't tested it)

```js
glob.sync('?(@*/)generator-*')
```

**Comment by SBoudrias** [comment]
Hey, thanks for starting to work on this feature!

We'll also need unit tests before this can be merged.

File: lib/resolver.js (hunk 3 above)
Comment (molant) [inline lib/resolver.js]: What about doing something like this:
```
var found = glob.sync('generator-*', { cwd: root, stat: true });
found = found.concat(glob.sync('@*/generator-*', { cwd: root, stat: true }));
```

I'm not a huge fun of complex regex (and I've tried several combinations and still haven't found the right pattern to match everything in one line). Would that be ok?

File: lib/environment.js (hunk 2 above)
Comment (molant) [inline lib/environment.js]: Not sure we want this exactly. There could potentially be some other folder up in the hierarchy with a @ and have nothing to do with the scope. That is why I was searching for the generator folder and then check if the father folder was a scope.
Does it make sense?
I'll make it more functional though.

File: lib/resolver.js (hunk 3 above)
Comment (SBoudrias) [inline lib/resolver.js]: Well, that would be better already.

cc @sindresorhus any idea for the glob pattern?

Comment (sindresorhus) [inline lib/resolver.js]: Use [`globby`](https://github.com/sindresorhus/globby) instead as it supports multiple patterns.

**Comment by molant** [comment]
@SBoudrias refactored the code and added a few tests. Let me know if I have to do something else.

File: test/resolver.js [hunk 4]
```diff
@@ -40,15 +57,18 @@ describe('Environment Resolver', function () {
     it('register local generators', function () {
       assert.ok(this.env.get('dummy:app'));
       assert.ok(this.env.get('dummy:yo'));
+      assert.ok(this.env.get('@dummyscope/scoped:app'));
     });
 
+    return;
```
Comment (SBoudrias) [inline test/resolver.js]: I think you forget this line

File: test/resolver.js [hunk 5]
```diff
@@ -40,15 +57,18 @@ describe('Environment Resolver', function () {
...
+
     it('register non-dependency local generator', function () {
       assert.ok(this.env.get('jquery:app'));
     });
 
     if (!process.env.NODE_PATH) {
       console.log('Skipping tests for global generators. Please setup `NODE_PATH` ' +
-        'environment variable to run it.');
+      'environment variable to run it.');
```
Comment (SBoudrias) [inline test/resolver.js]: Why did you change this line?

File: test/resolver.js [hunk 6]
```diff
@@ -68,7 +88,7 @@ describe('Environment Resolver', function () {
       before(function () {
         this.projectSubRoot = path.join(this.projectRoot, 'subdir');
         process.chdir(this.projectSubRoot);
-        shell.exec('npm install', { silent: true });
+        shell.exec('npm install', {silent: true});
```
Comment (SBoudrias) [inline test/resolver.js]: Why did you change this line?

File: lib/environment.js [hunk 7]
```diff
@@ -365,6 +365,11 @@ Environment.prototype.namespace = function namespace(filepath) {
     return ns.replace(lookup, '');
   }, ns);
 
+  var folders = ns.split(path.sep);
```
Comment (SBoudrias) [inline lib/environment.js]: move this code block together with the `if (scope)` statement

File: lib/environment.js [hunk 8]
```diff
@@ -373,6 +378,10 @@ Environment.prototype.namespace = function namespace(filepath) {
     .replace(/^[\/\\]+/, '') // remove leading `/`
     .replace(/[\/\\]+/g, ':'); // replace slashes by `:`
 
+  if (scope){
```
Comment (SBoudrias) [inline lib/environment.js]: Add a scape after the closing brace - e.g `if (scope) {`

File: test/resolver.js [hunk 9]
```diff
@@ -40,15 +57,18 @@ describe('Environment Resolver', function () {
     it('register local generators', function () {
       assert.ok(this.env.get('dummy:app'));
       assert.ok(this.env.get('dummy:yo'));
+      assert.ok(this.env.get('@dummyscope/scoped:app'));
```
Comment (SBoudrias) [inline test/resolver.js]: move this assertion to another `it()` block. "register generators in scoped packages"

**Comment by SBoudrias** [comment]
We'll need tests covering how the `namespace` method react when received scoped packages path.

File: test/resolver.js (hunk 5 above)
Comment (molant) [inline test/resolver.js]: Don't know what it got changed and I didn't see it on the diff tool :(

File: test/resolver.js (hunk 6 above)
Comment (molant) [inline test/resolver.js]: Don't know what it got changed and I didn't see it on the diff tool :(

File: lib/environment.js (hunk 7 above)
Comment (molant) [inline lib/environment.js]: I still need to keep a reference to the original value of ns at that moment. Do you want me to create another variable and then move that code inside the if or just move
```
var  scope = ...
```
inside the if?

File: test/resolver.js (hunk 6 above)
Comment (SBoudrias) [inline test/resolver.js]: Ok, I was wondering if it was from your editor, or if an update to our style checking forced you to update it to pass the test suite.

File: lib/environment.js (hunk 7 above)
Comment (SBoudrias) [inline lib/environment.js]: Ah no, I haven't realized this. Please overlook, it's fine the way it is.

**Comment by molant** [comment]
@SBoudrias let's try again!

**Comment by SBoudrias** [comment]
Code looks ok.

The lookup test seem brittle, but this test is to redo so I don't mind that much. But just know that using relative path is shooting in your foot as it'll cause test flakey-ness.

Travis is failing and it might be after a side effect of your changes. Can you look into it?

**Comment by molant** [comment]
@SBoudrias I've checked Travis' log and I don't know why it is failing. In the [current results](https://travis-ci.org/yeoman/environment/builds/46922774) I can see ESLint complaining about some part of the code I haven't modified with no-shadow (if I run it locally it works ok in all those files).
The rest of the errors/warning are the same as when I [previously passed the build](https://github.com/yeoman/environment/commit/4e72c7fce111f2e69f16a593fafe005bf8386061) and about the same as [build 35](https://travis-ci.org/yeoman/environment/builds/41138750) (with the same pending and failing results).

Running npm test on my branch and in the current master has the same ESLint errors.

Any idea of what could be wrong?

**Comment by SBoudrias** [comment]
Oh god, that's horrible... Our test suite was failing but travis was green... Sometime I really hate gulp.

I'll check it out as soon as I have a chance (your code LGTM, I'll merge when I'm done with that)

**Comment by SBoudrias** [comment]
Alright, so that's merged on master.

**Comment by molant** [comment]
Yipikaei! Thank you very much for all your feedback :)
